File: 366_find_leaves_of_binary_tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    // Time Complexity: O(log(n)) (traverse all nodes till leaves)
    // Space Complexity: O(n)

    // 特色：用 boolean 判斷是否為葉子，並移除掉該 leaf；同時保留非 leaf 的節點
    public List<List<Integer>> findLeaves(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        while (root != null) {
            List<Integer> temp = new ArrayList<>();
            // check root node is the leave
            if (isLeave(temp, root)) root = null;
            res.add(temp);
        }
        return res;
    }

    public boolean isLeave(List<Integer> temp, TreeNode node) {
        if (node != null && node.left == null && node.right == null) {
            temp.add(node.val);
            return true;
        }

        if (node != null && node.left != null) {
            // only remove leaf
            if (isLeave(temp, node.left)) node.left = null;
        }

        if (node != null && node.right != null) {
            // only remove leaf
            if (isLeave(temp, node.right)) node.right = null;
        }

        return false;
    }
}

File: python/366_find_leaves_of_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findLeaves(self, root: Optional[TreeNode]) -> List[List[int]]:
        def remove(node):
            if not node:
                return None
            if not node.left and not node.right:
                return None
            node.left = remove(node.left)
            node.right = remove(node.right)
            return node

        def findLeaves(node, res: List[TreeNode]):
            if not node: return
            if not node.left and not node.right:
                res.append(node.val)
            findLeaves(node.left, res)
            findLeaves(node.right, res)

        ans = []
        while root:
            leaves = []
            findLeaves(root, leaves)
            ans.append(list(leaves))
            root = remove(root)
        return ans


